pkg/controller/direct/mappings/validation.go

```diff
@@ -0,0 +1,251 @@
+							Proposal: proposal,
+						})
+						continue
+					}
+				}
+			}
+		} else {
+			for _, outField := range m.outType.Fields() {
+				id := outField.ID()
+
+				found := false
+				for _, mapping := range m.fields {
+					if len(mapping.OutPath.fields) == 1 && mapping.OutPath.fields[0] == id {
+						found = true
+					}
+				}
+				if found {
+					continue
+				}
+
+				ignore := false
+				for _, ignoreField := range m.ignore {
+					if ignoreField.ID == id {
+						ignore = true
+					}
+				}
+				if ignore {
+					continue
+				}
+				errors = append(errors, ValidationError{
+					Message: fmt.Sprintf("field %s is not mapped in %v", id, m.outType),
+					// Proposal: proposal,
+				})
+			}
+
+			for _, inField := range m.inType.Fields() {
+				id := inField.ID()
+
+				found := false
+				for _, mapping := range m.fields {
+					if len(mapping.InPath.fields) == 1 && mapping.InPath.fields[0] == id {
+						found = true
+					}
+				}
+				if found {
+					continue
+				}
+
+				ignore := false
+				for _, ignoreField := range m.ignore {
+					if ignoreField.ID == id {
+						ignore = true
+					}
+				}
+				if ignore {
+					continue
+				}
+				errors = append(errors, ValidationError{
+					Message: fmt.Sprintf("field %s is not mapped in %v", id, m.inType),
+				})
+			}
+		}
+
+	}
+
+	return errors
+}
+
+// buildGoField builds go code that could be used to add the relevant field.
+func buildGoField(f Field) string {
+	jsonName := f.getJSONKey()
+
+	fieldName := jsonToGoFieldName(jsonName)
+	fieldType := convertToGoType(f.Type().rt)
+	jsonTag := jsonName
+	jsonTag += ",omitempty"
+
+	requiredTag := ""
+	if f.isRequired() {
+		requiredTag = "true"
+	}
+
+	tags := []string{}
+	if jsonTag != "" {
+		tags = append(tags, fmt.Sprintf("json:%q", jsonTag))
+	}
+	if requiredTag != "" {
+		tags = append(tags, fmt.Sprintf("required:%q", requiredTag))
+	}
+
+	fieldTags := ""
+	if len(tags) != 0 {
+		fieldTags = " `" + strings.Join(tags, " ") + "`"
+	}
+
+	proposal := fmt.Sprintf("%s %s%s", fieldName, fieldType, fieldTags)
+	return proposal
+}
+
+// jsonToGoFieldName converts a json name into its go field name equivalent,
+// used to suggest go code.
+func jsonToGoFieldName(jsonName string) string {
+	var out []rune
+	for i, r := range jsonName {
+		if i == 0 {
+			r = unicode.ToUpper(r)
+		}
+		out = append(out, r)
+	}
+	return string(out)
+}
+
+// convertToGoType builds the go type name for the specified type,
+// used to suggest go code.
+func convertToGoType(t reflect.Type) string {
+	fieldGoType := t
+	switch fieldGoType.Kind() {
+	case reflect.Slice:
+		return "[]" + convertToGoType(t.Elem())
+	case reflect.Ptr:
+		return "*" + convertToGoType(t.Elem())
+	case reflect.Struct:
+		return t.Name()
+	case reflect.String:
+		return "string"
+	case reflect.Bool:
+		return "bool"
+	case reflect.Uint8:
+		return "uint8"
+	case reflect.Int32:
+		return "int32"
+	case reflect.Int64:
+		return "int64"
+	case reflect.Map:
+		return "map[todo]todo"
```

Comment:
Assuming this won't actually work. If true would it be better to fail here?

Reply:
We're outputing a proposed line to copy and paste into their go code.  We could output `"map[todo]todo /* TODO: replace with real types"` :-)  The problem with failing is that we don't give the caller anything to start from!